core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableSortedAggregate.java

```diff
@@ -0,0 +1,98 @@
+import org.apache.calcite.rel.RelCollationTraitDef;
+import org.apache.calcite.rel.RelCollations;
+import org.apache.calcite.rel.RelFieldCollation;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.AggregateCall;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.calcite.util.Pair;
+import org.apache.calcite.util.Util;
+import org.apache.calcite.util.mapping.Mappings;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Sort based physical implementation of {@link Aggregate} in
+ * {@link EnumerableConvention enumerable calling convention}. */
+public class EnumerableSortedAggregate extends Aggregate implements EnumerableRel {
+  public EnumerableSortedAggregate(
+      RelOptCluster cluster,
+      RelTraitSet traitSet,
+      RelNode input,
+      ImmutableBitSet groupSet,
+      List<ImmutableBitSet> groupSets,
+      List<AggregateCall> aggCalls) {
+    super(cluster, traitSet, ImmutableList.of(), input, groupSet, groupSets, aggCalls);
+    assert getConvention() instanceof EnumerableConvention;
+  }
+
+  @Override public EnumerableSortedAggregate copy(RelTraitSet traitSet, RelNode input,
+      ImmutableBitSet groupSet,
+      List<ImmutableBitSet> groupSets, List<AggregateCall> aggCalls) {
+    return new EnumerableSortedAggregate(getCluster(), traitSet, input,
+        groupSet, groupSets, aggCalls);
+  }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      final RelTraitSet required) {
+    if (!isSimple(this)) {
+      return null;
+    }
+
+    RelTraitSet inputTraits = getInput().getTraitSet();
+    RelCollation collation = required.getTrait(RelCollationTraitDef.INSTANCE);
+    ImmutableBitSet requiredKeys = ImmutableBitSet.of(RelCollations.ordinals(collation));
+    ImmutableBitSet groupKeys = ImmutableBitSet.range(groupSet.cardinality());
+
+    if (!groupKeys.contains(requiredKeys)) {
+      return null;
+    }
+
+    Mappings.TargetMapping mapping = Mappings.source(groupSet.toList(),
+        input.getRowType().getFieldCount());
+
+    if (requiredKeys.equals(groupKeys)) {
+      RelCollation inputCollation = RexUtil.apply(mapping, collation);
+      return Pair.of(required, ImmutableList.of(inputTraits.replace(inputCollation)));
+    } else if (groupKeys.contains(requiredKeys)) {
+      // group by a,b,c order by c,b
+      List<RelFieldCollation> list = new ArrayList<>(collation.getFieldCollations());
+      groupKeys.except(requiredKeys).forEach(k -> list.add(new RelFieldCollation(k)));
+      RelCollation aggCollation = RelCollations.of(list);
+      RelCollation inputCollation = RexUtil.apply(mapping, aggCollation);
+      return Pair.of(traitSet.replace(aggCollation),
+          ImmutableList.of(inputTraits.replace(inputCollation)));
+    }
+
+    // should not reach here
+    return null;
+  }
```

Comment:
If it should not be reached, how about throw an exception?

Reply:
done